<a href="https://colab.research.google.com/github/senisekoni32-a11y/ml_internship_tasks/blob/main/Task_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

names = ["CRIM", "ZN", "INDUS", "CHAS", "NOX", "RM", "AGE", "DIS",
         "RAD", "TAX", "PTRATIO", "B", "LSTAT", "MEDV"]

df = pd.read_csv("house_prediction_dataset.csv", sep=r"\s+", header=None, names=names)
print(df.shape)
df.head()

(506, 14)


,CRIM,ZN,INDUS,CHAS,NOX,RM,AGE,DIS,RAD,TAX,PTRATIO,B,LSTAT,MEDV
0,0.00632,18.0,2.31,0,0.538,6.575,65.2,4.0900,1,296.0,15.3,396.90,4.98,24.0
1,0.02731,0.0,7.07,0,0.469,6.421,78.9,4.9671,2,242.0,17.8,396.90,9.14,21.6
2,0.02729,0.0,7.07,0,0.469,7.185,61.1,4.9671,2,242.0,17.8,392.83,4.03,34.7
3,0.03237,0.0,2.18,0,0.458,6.998,45.8,6.0622,3,222.0,18.7,394.63,2.94,33.4
4,0.06905,0.0,2.18,0,0.458,7.147,54.2,6.0622,3,222.0,18.7,396.90,5.33,36.2


In [2]:
from sklearn.model_selection import train_test_split

df = df.drop("B", axis=1)          # dropped for the reason above

X = df.drop("MEDV", axis=1)        # clues: the neighbourhood descriptions
y = df["MEDV"]                     # answer: the median price

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print(X_train.shape, X_test.shape)

(404, 12) (102, 12)


In [5]:
from sklearn.linear_model import LinearRegression as lr

model = lr()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
print(y_pred[:5])

[28.95821495 36.28754087 14.09883571 25.03314354 17.60196806]


In [6]:
print(y_pred[:5])
print(y_test.values[:5])

[28.95821495 36.28754087 14.09883571 25.03314354 17.60196806]
[23.6 32.4 13.6 22.8 16.1]


In [7]:
from sklearn.metrics import mean_squared_error, r2_score

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print(mse, r2)

22.77770856386609 0.6893967884614753


In [8]:
coef = pd.Series(model.coef_, index=X.columns).sort_values()
print(coef)
print(model.intercept_)

NOX       -18.183246
DIS        -1.465507
PTRATIO    -0.894845
LSTAT      -0.537770
CRIM       -0.127196
TAX        -0.011275
AGE        -0.003594
INDUS       0.025248
ZN          0.030795
RAD         0.239990
CHAS        3.049477
RM          4.240098
dtype: float64
36.97046906703031


## Results

This linear regression model predicts median house prices for Boston-area neighbourhoods using 12 features (I dropped the `B` column because it is built from the racial makeup of each town). The data is from the 1970s, so prices are in 1970s dollars.

- **R² = 0.689:** the model explains about 69% of the variation in house prices on the test set.
- **Typical error ≈ $4,800** (the square root of the MSE of 22.8, which puts the error back into price units).

**What the coefficients show:** holding everything else fixed, each extra room per home (`RM`) adds about $4,240 to the predicted price, and each extra percentage point of lower-income residents (`LSTAT`) lowers it by about $540. These are patterns in the data and don't prove cause.